arangod/Cluster/Maintenance/Actions/UpgradeCollection.cpp

```diff
@@ -0,0 +1,766 @@
+        "failed to send and execute transaction to set shard upgrade status");
+  } else {
+    auto& feature = collection.vocbase().server().getFeature<arangodb::ClusterFeature>();
+    auto& ci = feature.clusterInfo();
+    auto& cache = feature.agencyCache();
+    std::uint64_t index = ::extractIndexFromTransactionResult(result);
+    cache.waitFor(index).wait();
+    ci.loadCurrent();
+  }
+}
+
+void removeStatusFromCurrent(arangodb::LogicalCollection& collection,
+                             arangodb::maintenance::ActionDescription const& desc) {
+  auto operations = std::vector<arangodb::AgencyOperation>{};
+
+  std::string const statusKey =
+      "/Current/Collections/" + desc.get(arangodb::maintenance::DATABASE) +
+      "/" + desc.get(arangodb::maintenance::COLLECTION) + "/" +
+      desc.get(arangodb::maintenance::SHARD) + "/" + arangodb::maintenance::UPGRADE_STATUS;
+  operations.emplace_back(statusKey, arangodb::AgencySimpleOperationType::DELETE_OP);
+
+  std::string const versionKey = "/Current/Version";
+  operations.emplace_back(versionKey, arangodb::AgencySimpleOperationType::INCREMENT_OP);
+
+  auto trx = arangodb::AgencyWriteTransaction{operations};
+
+  arangodb::AgencyComm comm(collection.vocbase().server());
+  arangodb::AgencyCommResult result = comm.sendTransactionWithFailover(trx);
+  if (!result.successful()) {
+    throw std::runtime_error(
+        "failed to send and execute transaction to set shard upgrade status");
+  } else {
+    auto& feature = collection.vocbase().server().getFeature<arangodb::ClusterFeature>();
+    auto& ci = feature.clusterInfo();
+    ci.loadCurrent();
+  }
+}
+
+std::unordered_map<::UpgradeState, std::vector<std::string>> serversByStatus(
+    arangodb::LogicalCollection& collection, std::unique_lock<std::mutex>& lock) {
+  TRI_ASSERT(lock.owns_lock());
+  std::unordered_map<::UpgradeState, std::vector<std::string>> map;
+
+  arangodb::LogicalCollection::UpgradeStatus& status = collection.upgradeStatus();
+  for (auto& it : status.map()) {
+    map[it.second].emplace_back(it.first);
+  }
+
+  return map;
+}
+
+void setStatusesForRollback(arangodb::LogicalCollection& collection,
+                            std::unique_lock<std::mutex>& lock) {
+  TRI_ASSERT(lock.owns_lock());
+  arangodb::LogicalCollection::UpgradeStatus& status = collection.upgradeStatus();
+  status.setError("");
+
+  auto& map = status.map();
+  for (auto const& [server, state] : map) {
+    if (state == ::UpgradeState::ToDo || state == ::UpgradeState::Prepare) {
+      // effectively already rolled back, so mark it as such
+      status.set(server, ::UpgradeState::Rollback);
+    }
+  }
+}
+}  // namespace
+
+namespace arangodb::maintenance {
+
+UpgradeCollection::UpgradeCollection(MaintenanceFeature& feature, ActionDescription const& desc)
+    : ActionBase(feature, desc), _timeout(600), _isSmartChild(false), _inRollback(false) {
+  std::stringstream error;
+
+  _labels.emplace(FAST_TRACK);
```

Comment:
This is definitely not a fast track job!

Reply:
agree